src/utils/allocators.zig

```diff
@@ -2,6 +2,269 @@ const std = @import("std");
+                .records = std.ArrayList(Record).init(allocator_config.records_allocator),
+                .memory = std.ArrayList([]T).init(allocator_config.records_allocator),
+                .memory_allocator = allocator_config.memory_allocator,
+                .capacity = 0,
+            };
+        }
+
+        pub fn create(allocator_config: AllocatorConfig) !*Self {
+            const self = try allocator_config.records_allocator.create(Self);
+            self.* = Self.init(allocator_config);
+            return self;
+        }
+
+        pub fn deinit(self: *Self) void {
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            for (self.memory.items) |block| {
+                self.memory_allocator.free(block);
+            }
+            self.memory.deinit();
+            self.records.deinit();
+        }
+
+        /// append a block of N elements to the manager
+        pub fn append(self: *Self, n: u64) std.mem.Allocator.Error!void {
+            if (n == 0) return;
+
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            try self.records.ensureUnusedCapacity(1);
+            try self.memory.ensureUnusedCapacity(1);
+
+            const buf = try self.memory_allocator.alloc(T, n);
+            // NOTE: we do this here so bincode serialization can work correctly
+            // otherwise, we run into undefined memory which breaks bincode
+            @memset(buf, default_init);
+
+            self.records.appendAssumeCapacity(.{
+                .is_free = true,
+                .buf = buf,
+                .global_index = self.capacity,
+                .len = buf.len,
+                .memory_index = self.memory.items.len,
+            });
+            self.memory.appendAssumeCapacity(buf);
+            self.capacity += buf.len;
+        }
+
+        const AllocError = error{
+            AllocTooBig,
+            AllocFailed,
+            AttemptedZeroAlloc,
+            // NOTE: even though this doesnt get hit on `alloc`, zig isnt smart enough to know that
+            CollapseFailed,
+        } || std.mem.Allocator.Error;
+
+        pub fn alloc(self: *Self, n: u64) AllocError!struct { []T, u64 } {
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            if (n == 0) return error.AttemptedZeroAlloc;
+
+            for (0..config.max_collapse_tries) |_| {
+                return self.allocUnsafe(n) catch |err| {
+                    switch (err) {
+                        error.CollapseFailed => {
+                            if (config.thread_safe) self.mux.unlock();
+                            defer if (config.thread_safe) self.mux.lock();
+                            std.time.sleep(std.time.ns_per_ms * config.collapse_sleep_ms);
+                            // try again -- will hit collapse again
```

Comment:
for my understanding, looking at the implementation of `collapseUnsafe` I do not see what will make it succeed on retry so not sure why the logic here is to retry after some time

Reply:
if collapse fails then there wasnt enough free records which could collapse to support the size requested - if we wait some time and try to collapse again there may be _new_ free records (which were free'd by some other consumer thread) which can be collapsed and have the alloc call succeed

Reply:
added a comment - 0173e4d8